### PR TITLE
Allow registering same operator schema multiple times

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17779 Specify overload name in function schema&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14293878/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17780 Allow registering same operator schema multiple times**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14316791/)

Now that we have named overloads, we can allow registering the same function schema multiple times and just check it's identical.

This is going to be used in custom op registration since they register the schema every time a kernel is registered.

Differential Revision: [D14316791](https://our.internmc.facebook.com/intern/diff/D14316791/)

### CODE REVIEW

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 1]
```diff
@@ -60,10 +75,14 @@ void Dispatcher::deregisterSchema(const OperatorHandle& op) {
     AT_ERROR("Tried to deregister op schema that still has kernels registered");
   }
 
-  // note: call listeners *before* operator is removed, i.e. dispatcher is still valid for removed op
-  listeners_->callOnOperatorDeregistered(op);
+  // reduce refcount and actually deregister if no references left
+  --op.operatorDefIterator_->refcount;
+  if (0 == --op.operatorDefIterator_->refcount) {
```
Comment (dzhulgakov) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp]: don't you decrease it twice? O_O

in general, would be nice to have a test for it. That being said, as I mentioned earlier, I'm not sure we'd practically use "unloading" part of operators

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -960,6 +960,21 @@ inline bool IValue::isSameIdentity(IValue& rhs) {
         && this->payload.as_intrusive_ptr == rhs.payload.as_intrusive_ptr;
   }
 }
+
+inline bool shallowEquals(const IValue& lhs, const IValue& rhs) {
```
Comment (dzhulgakov) [inline aten/src/ATen/core/ivalue.h]: why do we need shallow version? just to avoid comparing tensors?

**Review by dzhulgakov — COMMENTED** [review]
Given that you support "runtime" registration, it should be possible to write a unittest for dedup working

File: aten/src/ATen/core/ivalue.h (hunk 2 above)
Comment (smessmer) [reply aten/src/ATen/core/ivalue.h]: I'm not 100% confident how we want to define equality for IValues. We could make it shallow or we could make it deep and compare tensor contents, or anything inbetween. Note that deep comparison would be much harder for non-tensor Blobs with custom types.
So I took the easy way out and didn't try to define equality for IValues, but just defined a shallow equality, which is enough for my use case.

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp]: Yes, this already blew up when I tried to use it and it's fixed in the newer version. Unit tests are in the pipeline.

**Comment by smessmer** [comment]
closing because of ghexport issues

**Pull request closed without merging**